---
title: "Matt Kent RNAseq experiment"
author: "Matt Cannon"
date: "10-15-2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
    echo = TRUE,
    cache = TRUE,
    fig.height = 10,
    fig.width = 20,
    dpi = 300,
    bootstrap.show.code = FALSE,
    bootstrap.show.output = FALSE,
    bootstrap.show.message = FALSE
)
```



# RNAseq analysis of 

## Goals:

placeholder

## Samples:

placeholder

## Approach

placeholder

# Workflow

## Load libraries
```{r libraries, cache=FALSE, eval=TRUE}
library(tidyverse)
theme_set(theme_bw())
theme_update(plot.title = element_text(hjust = 0.5))
library(gt)
library(edgeR)
```

## Make up directory structure
```{bash mkdirs, eval=TRUE}
for directoryName in \
  misc \
  slurmOut \
  sbatchCmds \
  output/fastqc \
  output/figures \
  output/ERCC/aligned/ \
  output/geneCounts \
  output/counts 
do
  if [ ! -d ${directoryName} ]
  then
    mkdir -p ${directoryName}
  fi 
done
```

## Run fastqc to assess data quality
```{bash fastqc, eval=FALSE}
sbatch sbatchCmds/fastqc.sh

rm output/fastqc/*.zip

for inFile in output/fastqc/*/fastqc_data.txt
do
  perl ~/scripts/splitFastqcData.pl -i ${inFile}
done
```


### Plot Fastqc output for all samples
```{r plotFastqc, eval=TRUE}
adapter_file_dirs <- list.dirs(
    path = "output/fastqc",
    full.names = TRUE,
    recursive = FALSE
)

to_plot <- tibble(
    file = c(
        "Adapter_Content.txt",
        "Per_base_sequence_quality.txt"
    ),
    column = c(
        "Nextera_Transposase_Sequence",
        "Mean"
    )
)

for (i in 1:nrow(to_plot)) {
    temp_data <- NULL

    for (data_dir in adapter_file_dirs) {
        temp_data <- read_delim(paste(data_dir,
            "/split_data/",
            to_plot$file[i],
            sep = ""
        ),
        delim = "\t",
        col_names = TRUE,
        show_col_types = FALSE
        ) %>%
            rename_all(~ str_replace_all(., " ", "_")) %>%
            rename_all(~ str_remove(., "[#']")) %>%
            select(Base, to_plot$column[i]) %>%
            mutate(Sample = str_remove(data_dir, "_001_fastqc") %>%
                str_remove(".+\\/")) %>%
            mutate(Read = str_remove(Sample, ".+_")) %>%
            mutate(Tech = str_remove(Sample, "_.+")) %>%
            bind_rows(temp_data)
    }

    max_y <- ceiling(max(temp_data[[to_plot$column[i]]]) * 1.1)

    plot_fastqc <- ggplot(temp_data, aes(
        x = Base,
        y = get(to_plot$column[i]),
        group = Sample,
        color = Sample
    )) +
        geom_line() +
        geom_point() +
        facet_wrap(~Tech, ncol = 1) +
        ylim(0, max_y) +
        ylab(to_plot$column[i]) +
        theme(legend.position = "none") +
        ggtitle(str_remove(to_plot$file[i], ".txt"))

    png(
        filename = paste("output/figures/fastqc_",
            str_remove(to_plot$file[i], ".txt"),
            ".png",
            sep = ""
        ),
        width = 7000,
        height = 3000,
        res = 300
    )
    print(plot_fastqc)
    dev.off()
}

input_reads <- read_tsv(list.files(
    path = "output/fastqc",
    recursive = TRUE,
    pattern = "Basic_Statistics.txt",
    full.names = TRUE
),
id = "File",
show_col_types = FALSE
) %>%
    filter(`#Measure` == "Total Sequences") %>%
    select(-`#Measure`) %>%
    mutate(
        File = str_remove(File, "output/fastqc/") %>%
            str_remove("_001_fastqc/split_data/Basic_Statistics.txt"),
        Sample = str_remove(File, "_S[0-9].+"),
        Lane = str_remove(File, ".+_L00") %>%
            str_remove("_R[12]"),
        Read = str_replace(File, ".+_R", "R"),
        Value = as.numeric(Value)
    ) %>%
    rename(Reads = Value)

ggplot(input_reads, aes(x = Sample, y = Reads / 1000000, fill = Lane)) +
    geom_col() +
    facet_wrap(~Read, ncol = 1) +
    ylab("Reads (in millions)") +
    xlab("") +
    ggtitle("Sequenced Reads Per Sample") +
    theme(axis.text.x = element_text(angle = 90, hjust = 1))

ggsave("output/figures/InputReads.png",
    height = 3000,
    width = 4000,
    units = "px"
)
```

## Align to zebrafish genome
Reference includes the ERCC reference RNA sequences

```{bash align, eval=FALSE}
sbatch sbatchCmds/align.sh
```

### Plot percent of reads aligned
```{r plotAlignedPercent, dependson='align', eval=TRUE}
aligned_percent <- system("grep 'overall alignment rate' output/aligned/*Summary.txt",
    intern = TRUE
) %>%
    tibble() %>%
    rename(File = ".") %>%
    separate(File,
        into = c("Sample", "Alignment"),
        sep = ":",
        remove = TRUE
    ) %>%
    mutate(
        Sample = str_remove(Sample, "output/aligned/") %>%
            str_remove("Summary.txt"),
        Alignment = str_remove(Alignment, "%.+") %>%
            as.numeric()
    )

ggplot(aligned_percent, aes(x = Sample, y = Alignment)) +
    geom_col() +
    labs(
        x = "",
        y = "Percent Aligned",
        title = "Percent of Input Reads Aligned to the Genome"
    ) +
    theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
    scale_y_continuous(n.breaks = 10, limits = c(0, 100))

ggsave("output/figures/PercentAligned.png",
    height = 3000,
    width = 4000,
    units = "px"
)
```

## Count number of reads per gene
Annotation includes the ERCC references
```{bash countGenes, dependson='align', eval=FALSE}
sbatch sbatchCmds/countGenes.sh
```

### Plot the amount of reads assigned to genes
```{r plotGeneAssignment, dependson='countGenes', eval=TRUE}
genic_reads <- read_tsv("output/geneCounts/combined.txt.summary",
    comment = "#",
    show_col_types = FALSE
) %>%
    rename_with(~ str_remove(.x, "output/aligned/")) %>%
    rename_with(~ str_remove(.x, ".bam")) %>%
    pivot_longer(-Status,
        names_to = "Sample",
        values_to = "Reads"
    ) %>%
    filter(Status == "Assigned" |
        Status == "Unassigned_NoFeatures" |
        Status == "Unassigned_Ambiguity") %>%
    mutate(Status = str_replace(
        Status,
        "Unassigned_NoFeatures",
        "Non-genic"
    ) %>%
        str_replace("Unassigned_Ambiguity", "Multiple Features")) %>%
    group_by(Sample) %>%
    mutate(Percent_Reads = Reads / sum(Reads) * 100)


ggplot(genic_reads, aes(x = Sample, y = Percent_Reads, fill = Status)) +
    geom_col() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
    scale_fill_discrete(name = "Read Context") +
    xlab("") +
    ylab("Percent of reads for each sample") +
    ggtitle("Assignment of Sequencing Reads to Genes")

ggsave("output/figures/geneAssignmentPlot.png",
    width = 3000,
    height = 2000,
    units = "px",
    device = "png"
)
```

## Plot reads lost along the way 
```{r plotReadsLostAlongTheWay, eval=TRUE}
input_reads %>%
    filter(Read == "R1") %>%
    group_by(Sample) %>%
    summarize(Sequenced_Reads = sum(Reads)) %>%
    full_join(aligned_percent, by = "Sample") %>%
    mutate(Aligned_Reads = (Alignment / 100) * Sequenced_Reads %>%
        round()) %>%
    full_join(genic_reads %>%
        filter(Status == "Assigned") %>%
        rename(Genic_Reads = Reads) %>%
        select(Sample, Genic_Reads),
    by = "Sample"
    ) %>%
    select(-Alignment) %>%
    pivot_longer(-Sample,
        names_to = "Stage",
        values_to = "Reads"
    ) %>%
    mutate(Stage = fct_relevel(
        Stage,
        "Sequenced_Reads",
        "Aligned_Reads",
        "Genic_Reads"
    )) %>%
    ggplot(aes(x = Sample, y = Reads / 1000000, fill = Stage)) +
    geom_col(position = "dodge") +
    ylab("Reads (in millions)") +
    xlab("") +
    theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
    ggtitle("Number of Reads Per Sample at Each Stage")

ggsave("output/figures/ReadsLostAlongTheWay.png",
    width = 3000,
    height = 2000,
    units = "px",
    device = "png"
)
```


## Plot Her3 expression
```{r her3Expression, eval=TRUE}
read_tsv("output/geneCounts/combined.txt",
    show_col_types = FALSE,
    comment = "#"
) %>%
    filter(Geneid == "ENSDARG00000076857.4") %>%
    select(starts_with("output")) %>%
    rename_with(~ str_remove(.x, "output/aligned/")) %>%
    rename_with(~ str_remove(.x, ".bam")) %>%
    pivot_longer(
        cols = everything(),
        names_to = "Sample",
        values_to = "Her3"
    ) %>%
    mutate(Group = str_replace(Sample, "_[123]_", "_")) %>%
    ggplot(aes(
        x = Group,
        y = Her3,
        fill = Group
    )) +
    geom_boxplot() +
    geom_point(color = "black") +
    theme(
        axis.text.x = element_text(angle = 90, hjust = 1),
        legend.position = "none"
    )

ggsave("output/figures/unNormHer3.png",
    width = 3000,
    height = 2000,
    units = "px",
    device = "png"
)
```

## Pairwise plots of un-normalized data
```{r pairwisePlots, eval=TRUE}
read_tsv("output/geneCounts/combined.txt",
    show_col_types = FALSE,
    comment = "#"
) %>%
    select(starts_with("output")) %>%
    rename_with(~ str_remove(.x, "output/aligned/")) %>%
    rename_with(~ str_remove(.x, ".bam")) %>%
    mutate(gene_sum = rowSums(across())) %>%
    filter(gene_sum > 10) %>%
    select(-gene_sum) %>%
    GGally::ggpairs()

ggsave("output/figures/unNormPairs.png",
    width = 8000,
    height = 8000,
    units = "px",
    device = "png"
)
```

## Use edgeR to test for differential expression
```{r edgeR, eval=TRUE}
group_table <- tibble(samples = list.files(
    path = "output/aligned",
    pattern = ".bam$"
)) %>%
    mutate(samples = str_remove(samples, "output/aligned/") %>%
        str_remove(".bam")) %>%
    mutate(group = str_replace(samples, "_[123]_", "_"))

group_list <- group_table %>%
    select(group) %>%
    distinct() %>%
    pull(group)

gene_names <- read_tsv("/gpfs0/home1/gdlessnicklab/mvc002/analyses/kendall/refs/zebrafish_gene_info.txt.gz",
    show_col_types = FALSE
) %>%
    rename(Gene = `Gene stable ID version`) %>%
    select(
        -`Transcript stable ID version`,
        -`Transcript stable ID`,
        -`Transcript type`,
        -`Gene type`,
        -`Gene stable ID`
    ) %>%
    mutate(Gene = str_remove(Gene, "\\.[0-9]+")) %>%
    unique() %>%
    rename(
        gene_name = `Gene name`
    )

# Loop through all combinations of groups
for (i in seq_len(length(group_list) - 1)) {
    for (j in (i + 1):length(group_list)) {
        group_1 <- group_list[i]
        group_2 <- group_list[j]

        # Get the samples in each group
        samples <- group_table %>%
            filter(group == group_1 | group == group_2) %>%
            select(samples) %>%
            pull(samples)

        groups <- group_table %>%
            filter(group == group_1 | group == group_2) %>%
            select(group) %>%
            pull(group)

        gene_expr <- read_tsv("output/geneCounts/combined.txt",
            col_names = TRUE,
            comment = "#",
            show_col_types = FALSE
        ) %>%
            select(-Chr, -Start, -End, -Strand, -Length) %>%
            rename_all(~ str_remove(., "output/aligned/")) %>%
            rename_all(~ str_remove(., "_.bam")) %>%
            rename_all(~ str_remove(., ".R.bam")) %>%
            rename_all(~ str_remove(., ".bam")) %>%
            rename_all(~ str_remove(., ".+/")) %>%
            select(all_of(samples), Geneid) %>%
            pivot_longer(
                cols = !Geneid,
                names_to = "Sample",
                values_to = "Reads"
            ) %>%
            mutate(Geneid = str_remove(Geneid, "\\.[0-9]+")) %>%
            group_by(Geneid, Sample) %>%
            summarize(Reads = sum(Reads)) %>%
            ungroup() %>%
            pivot_wider(names_from = Sample, values_from = Reads) %>%
            column_to_rownames(var = "Geneid")

        ERCC_genes <- gene_expr %>%
            rownames_to_column("gene") %>%
            as_tibble() %>%
            filter(grepl("ERCC", gene)) %>%
            column_to_rownames(var = "gene")

        gene_expr <- gene_expr %>%
            rownames_to_column("gene") %>%
            filter(!grepl("ERCC", gene)) %>%
            column_to_rownames(var = "gene")

        data_dge <- DGEList(counts = gene_expr, group = groups)

        kept_df <- filterByExpr(data_dge, min.count = 10)

        data_dge <- data_dge[kept_df, keep.lib.sizes = FALSE] %>%
            calcNormFactors(.) %>%
            estimateDisp(., model.matrix(~groups))

        dim(data_dge$counts)

        tested <- exactTest(data_dge)

        results <- topTags(tested, n = Inf)
        results$table$Gene <- rownames(results$table)

        summarized_data <- cpm(data_dge) %>%
            as.data.frame() %>%
            rownames_to_column(var = "Gene") %>%
            as_tibble() %>%
            full_join(cpmByGroup(data_dge) %>%
                as.data.frame() %>%
                rename_all(~ str_c(.x, "_mean")) %>%
                rownames_to_column("Gene") %>%
                as_tibble(),
            by = "Gene"
            ) %>%
            full_join(results$table, by = "Gene") %>%
            left_join(gene_names, by = "Gene")


        write.table(summarized_data,
            file = paste("output/edgeROut/edgeR_",
                group_1,
                "_",
                group_2,
                ".txt",
                sep = ""
            ),
            quote = FALSE,
            sep = "\t",
            row.names = FALSE
        )
        # Plotting ########################
        paste_end <- function(x) {
            paste(x, "_", group_1, "_", group_2, ".png", sep = "")
        }

        ### logFC histogram
        print(ggplot(summarized_data, aes(x = logFC)) +
            geom_vline(xintercept = 0, color = "red") +
            geom_histogram(bins = 100) +
            ggtitle(paste("Histogram of logFC\n",
                group_1,
                "\n",
                group_2,
                sep = ""
            )))

        ggsave(filename = paste_end("output/figures/edgeR/logFCHistogram"),
            width = 2000,
            height = 2000,
            units = "px",
            device = "png"
        )

        print(ggplot(summarized_data, aes(x = PValue)) +
            geom_histogram(bins = 100) +
            ggtitle(paste("Histogram of p-values\n",
                group_1,
                "\n",
                group_2,
                sep = ""
            )))


        ggsave(paste_end("output/figures/edgeR/pvalueHistogram"),
            width = 2000,
            height = 2000,
            units = "px",
            device = "png"
        )

        print(ggplot(summarized_data, aes(x = FDR)) +
            geom_histogram(bins = 100) +
            ggtitle(paste("Histogram of FDR values\n",
                group_1,
                "\n",
                group_2,
                sep = ""
            )))

        ggsave(paste_end("output/figures/edgeR/fdrHistogram"),
            width = 2000,
            height = 2000,
            units = "px",
            device = "png"
        )

        print(ggplot(summarized_data, aes(
            x = logFC,
            y = -log10(PValue),
            color = FDR <= 0.1
        )) +
            geom_point(alpha = 0.5) +
            ggtitle(paste("Volcano plot\nFDR <= 0.1 denoted in blue\n",
                group_1,
                "\n",
                group_2,
                sep = ""
            )))

        ggsave(paste_end("output/figures/edgeR/volcanoPlot"),
            width = 2000,
            height = 2000,
            units = "px",
            device = "png"
        )

        print(ggplot(
            summarized_data,
            aes(
                x = !! sym(paste(group_1, "_mean", sep = "")),
                y = !! sym(paste(group_2, "_mean", sep = ""))
            )) +
            geom_point(alpha = 0.2) +
            geom_abline(slope = 1, intercept = 0) +
            ggtitle(paste("Scatterplot of mean CPM values\n",
                group_1,
                "\n",
                group_2,
                sep = ""
            )) +
            scale_y_log10() +
            scale_x_log10())

        ggsave(paste_end("output/figures/edgeR/logGroupScatter"),
            width = 2000,
            height = 2000,
            units = "px",
            device = "png"
        )

        print(ggplot(
            summarized_data,
            aes(
                x = !! sym(paste(group_1, "_mean", sep = "")),
                y = !! sym(paste(group_2, "_mean", sep = "")
            )
            )
        ) +
            geom_point(alpha = 0.2) +
            geom_abline(slope = 1, intercept = 0) +
            ggtitle(paste("Scatterplot of mean CPM values\n",
                group_1,
                " ",
                group_2,
                sep = ""
            )))

        ggsave(paste_end("output/figures/edgeR/groupScatter"),
            width = 2000,
            height = 2000,
            units = "px",
            device = "png"
        )
    }
}
```

## PCA analysis
```{r pca, eval=TRUE}
group_table <- tibble(samples = list.files(
    path = "output/aligned",
    pattern = ".bam$"
)) %>%
    mutate(samples = str_remove(samples, "output/aligned/") %>%
        str_remove(".bam")) %>%
    mutate(group = str_replace(samples, "_[123]_", "_"))

group_list <- group_table %>%
    select(group) %>%
    distinct() %>%
    pull(group)

gene_expr <- read_tsv("output/geneCounts/combined.txt",
            col_names = TRUE,
            comment = "#",
            show_col_types = FALSE
        ) %>%
            select(-Chr, -Start, -End, -Strand, -Length) %>%
            rename_all(~ str_remove(., "output/aligned/")) %>%
            rename_all(~ str_remove(., "_.bam")) %>%
            rename_all(~ str_remove(., ".R.bam")) %>%
            rename_all(~ str_remove(., ".bam")) %>%
            rename_all(~ str_remove(., ".+/")) %>%
            pivot_longer(
                cols = !Geneid,
                names_to = "Sample",
                values_to = "Reads"
            ) %>%
            mutate(Geneid = str_remove(Geneid, "\\.[0-9]+")) %>%
            group_by(Geneid, Sample) %>%
            summarize(Reads = sum(Reads)) %>%
            ungroup() %>%
            pivot_wider(names_from = Sample, values_from = Reads) %>%
            column_to_rownames(var = "Geneid")

data_dge <- DGEList(counts = gene_expr, group = group_table$group)

kept_df <- filterByExpr(data_dge, min.count = 10)

data_dge <- data_dge[kept_df, keep.lib.sizes = FALSE] %>%
    calcNormFactors(.)

dim(data_dge$counts)

gene_counts <- cpm(data_dge) %>%
    as.data.frame() %>%
    t()

pca <- prcomp(gene_counts)

variance <- (pca$sdev)^2 / sum(pca$sdev^2) * 100
names(variance) <- colnames(pca$x)

plot(variance)

prin_comps <- pca$x %>%
    as.data.frame() %>%
    rownames_to_column(var = "samples") %>%
    as_tibble() %>%
    full_join(group_table)

ggplot(prin_comps, aes(x = PC1, y = PC2, color = group)) +
    geom_point() +
    ggtitle("PCA Plot for all Samples") +
    ggrepel::geom_text_repel(aes(label = samples),
                             size = 2) +
    xlab(paste("PC1 (",
                round(variance[1], 1),
                "%)",
                sep = "")) +
    ylab(paste("PC2 (",
        round(variance[2], 1),
        "%)",
        sep = ""
    )) +
    theme(legend.position = "none")

ggsave("output/figures/allSamplePCA.png",
    width = 2500,
    height = 2000,
    units = "px",
    device = "png"
)
png
for (timepoint in c("24hpf", "72hpf")) {

    sub_pca <- gene_counts %>%
        as.data.frame() %>%
        rownames_to_column(var = "samples") %>%
        filter(grepl(timepoint, samples)) %>%
        column_to_rownames(var = "samples") %>%
        prcomp()

    sub_prin_comps <- sub_pca$x %>%
        as.data.frame() %>%
        rownames_to_column(var = "samples") %>%
        as_tibble() %>%
        full_join(group_table)

    variance <- (sub_pca$sdev)^2 / sum(sub_pca$sdev^2) * 100
    names(variance) <- colnames(sub_pca$x)
    plot(variance)

    ggplot(sub_prin_comps, aes(x = PC1, y = PC2, color = group)) +
        geom_point() +
        ggtitle(paste("PCA Plot for", timepoint, "Samples")) +
        ggrepel::geom_text_repel(aes(label = samples),
                                 size = 3) +
        xlab(paste("PC1 (",
                   round(variance[1], 1),
                   "%)",
                   sep = "")) +
        ylab(paste("PC2 (",
            round(variance[2], 1),
            "%)",
            sep = ""
        )) +
        theme(legend.position = "none")

    ggsave(paste("output/figures/",
                 timepoint,
                 "PCA.png",
                 sep = ""),
        width = 2500,
        height = 2000,
        units = "px",
        device = "png"
    )
}
```
# Conclusions

placeholder
 
 
snp calling
